### PR TITLE
Added activation offload

### PR DESCRIPTION
**Pull request by Gy-Lu**
Have passed the unit test.

### CODE REVIEW

File: tests/test_utils/test_activation_checkpointing.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def forward(x, weight):
 
 
 @pytest.mark.gpu
```
Comment (feifeibear) [inline tests/test_utils/test_activation_checkpointing.py:21]: @pytest.mark.parametrize(cpu_offload [True, False])
def test_activation_checkpointing(cpu_offload):

File: colossalai/utils/activation_checkpoint.py [hunk 2]
```diff
@@ -70,8 +77,9 @@ def backward(ctx, *args):
 
         # Fill in inputs with appropriate saved tensors.
         for i, idx in enumerate(tensor_indices):
-            inputs[idx] = tensors[i]
-
+            tmp = tensors[i].detach().to(ctx.device)
+            tmp.requires_grad = tensors[i].requires_grad
```
Comment (feifeibear) [inline colossalai/utils/activation_checkpoint.py:81]: Both detach() and requires_grad setting are unnecessary.
Because inputs will be detached later in detach_variable.

File: colossalai/utils/activation_checkpoint.py [hunk 3]
```diff
@@ -32,7 +34,12 @@ def forward(ctx, run_function, *args):
         tensor_inputs = []
```
Comment (feifeibear) [inline colossalai/utils/activation_checkpoint.py:34]: You missed an import procedure.
The forward can be executed during BWD of the model (grad checkpointing).
args should be guaranteed on cuda device before passing to run_function(*args).
I believe in some cases the args may resident on CPU.
https://github.com/microsoft/DeepSpeed/blob/013df6a9e8aef5d20ae0cf059c5fbd81cf90b694/deepspeed/runtime/activation_checkpointing/checkpointing.py#L569

File: tests/test_utils/test_activation_checkpointing.py (hunk 1 above)
Comment (Gy-Lu) [reply tests/test_utils/test_activation_checkpointing.py:21]: For we use  `add_seed` in the function. It can not be re-run to test.
Have fixed it and added the pytest.

File: colossalai/utils/activation_checkpoint.py [hunk 4]
```diff
@@ -5,14 +5,16 @@
 from torch.utils.checkpoint import check_backward_validity, detach_variable
 
 from colossalai.context.random import get_states, get_current_mode, set_seed_states, set_mode, sync_states
-
+from .cuda import get_current_device
```
[the remark this answers is not in the document]
Comment (Gy-Lu) [reply colossalai/utils/activation_checkpoint.py:8]: It will cause an ImportError.
For the `activation_checkpoint.py` and `cuda.py` are both under `utils`, and used by the same `__init__`.

File: colossalai/utils/activation_checkpoint.py (hunk 2 above)
Comment (Gy-Lu) [reply colossalai/utils/activation_checkpoint.py:81]: The `detach()` seems to be unnecessary, for `to()` would create a new tensor.
But the `requires_grad` is necessary, I have tested that the to() method would set `requires_grad=False` as well.

Comment (Gy-Lu) [reply colossalai/utils/activation_checkpoint.py:81]: The `detach()` seems to be unnecessary, for `to()` would create a new tensor.
But the `requires_grad` is necessary, I have tested that the to() method would set `requires_grad=False` as well.

File: colossalai/utils/activation_checkpoint.py (hunk 3 above)
Comment (Gy-Lu) [reply colossalai/utils/activation_checkpoint.py:34]: ok, I will try to add the strategy for this case

**Comment by feifeibear** [comment]
https://github.com/Oldpan/Pytorch-Memory-Utils
Can you use this tool to check current tensors in CPU and GPU.

**Comment by Gy-Lu** [comment]
> https://github.com/Oldpan/Pytorch-Memory-Utils Can you use this tool to check current tensors in CPU and GPU.

Hi, sorry for the late reply. I have used this tool and found that all tensors were in their positions.

**Review by feifeibear — APPROVED** [review]
